File: frontend/src/api/execute.ts
export interface ExecutionResult {
  stdout: string
  charts: string[]
  error: string | null
  execution_time_ms: number
}

export async function executeCode(params: {
  code: string
  mode: "quick" | "full"
  device: "cpu" | "mps"
}): Promise<ExecutionResult> {
  const response = await fetch("/api/execute", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(params),
  })

  if (!response.ok) {
    const text = await response.text().catch(() => "Unknown error")
    throw new Error(`Execution failed (${response.status}): ${text}`)
  }

  return response.json()
}

File: backend/app/routes/execute.py
from typing import Literal

from fastapi import APIRouter
from pydantic import BaseModel, Field

from app.executor import execute_code
from app.executor.models import ExecutionResult

router = APIRouter(prefix="/api")


class ExecuteRequest(BaseModel):
    code: str = Field(max_length=10_000)
    mode: Literal["quick", "full"]
    device: Literal["cpu", "mps"] = "cpu"


class ExecuteResponse(BaseModel):
    stdout: str
    charts: list[str]
    error: str | None
    execution_time_ms: float


@router.post("/execute", response_model=ExecuteResponse)
async def execute(request: ExecuteRequest) -> ExecuteResponse:
    result: ExecutionResult = await execute_code(
        code=request.code,
        mode=request.mode,
        device=request.device,
    )
    return ExecuteResponse(
        stdout=result.stdout,
        charts=result.charts,
        error=result.error,
        execution_time_ms=result.execution_time_ms,
    )
